common/src/main/java/io/netty/util/internal/Murmur3.java

```diff
@@ -0,0 +1,464 @@
+    static final Murmur3 INSTANCE_SAFE;
+    static final Murmur3 INSTANCE_SAFE_CASE_INSENSITIVE;
+    private final NativeTypeAccessor accessor;
+    private final int seed;
+
+    static {
+        Field stringValueField = AccessController.doPrivileged(new PrivilegedAction<Field>() {
+            @Override
+            public Field run() {
+                try {
+                    Field f = String.class.getDeclaredField("value");
+                    f.setAccessible(true);
+                    return f;
+                } catch (NoSuchFieldException e) {
+                    logger.warn("Failed to find String value array." +
+                            "String hash code optimizations are disabled.", e);
+                } catch (SecurityException e) {
+                    logger.info("No permissions to get String value array." +
+                            "String hash code optimizations are disabled.", e);
+                }
+                return null;
+            }
+        });
+        Field stringBuilderValueField = AccessController.doPrivileged(new PrivilegedAction<Field>() {
+            @Override
+            public Field run() {
+                try {
+                    Field f = StringBuilder.class.getSuperclass().getDeclaredField("value");
+                    f.setAccessible(true);
+                    return f;
+                } catch (NoSuchFieldException e) {
+                    logger.warn("Failed to find StringBuilder value array." +
+                            " Hash code optimizations for this type are disabled.", e);
+                } catch (SecurityException e) {
+                    logger.info("No permissions to get StringBuilder value array." +
+                            " Hash code optimizations for this type are disabled.", e);
+                }
+                return null;
+            }
+        });
+        Murmur3 unsafe = null;
+        Murmur3 unsafeAsciiLowerCase = null;
+        long stringValueFieldOffset = -1;
+        long stringBuilderValueFieldOffset = -1;
+        if (PlatformDependent0.hasUnsafe()) {
+            if (stringValueField != null) {
+                stringValueFieldOffset = PlatformDependent0.objectFieldOffset(stringValueField);
+            }
+            if (stringBuilderValueField != null) {
+                stringBuilderValueFieldOffset = PlatformDependent0.objectFieldOffset(stringBuilderValueField);
+            }
+
+            unsafe = new Murmur3(new AbstractNativeTypeAccessor() {
+                @Override
+                public int getInt(byte[] data, long i) {
+                    return PlatformDependent0.getInt(data, i);
+                }
+                @Override
+                public int getIntAsBytes(char[] data, long i) {
+                    long v = PlatformDependent0.getLong(data, CHAR_ARRAY_BASE_OFFSET + i * CHAR_ARRAY_INDEX_SCALE);
+                    return (int) ((v & 0xFF) |
+                            ((v >>> 8) & 0xFF00) |
```

Comment:
Not needed when masking. More used out of habit here.  I much more rarely need the sign extension in the shift, and that version is also conditional based upon the value where `>>>` always shifts a 0.  So if I don't need sign extension I tend to use `>>>`.